//
// PROJECT:         Aspia Remote Desktop
// FILE:            proto/desktop_session_message.proto
// LICENSE:         See top-level directory
// PROGRAMMERS:     Dmitry Chapyshev (dmitry@aspia.ru)
//

syntax = "proto3";

option optimize_for = LITE_RUNTIME;

package aspia.proto;

message KeyEvent
{
    // The keyboard (Caps/Num) lock states.
    enum Flags
    {
        UNKNOWN  = 0;
        CAPSLOCK = 1;
        NUMLOCK  = 2;
        EXTENDED = 4;
        PRESSED  = 8;
    }

    uint32 keycode = 1; // Windows virtual key code.
    uint32 flags = 2;
}

message PointerEvent
{
    enum ButtonMask
    {
        EMPTY         = 0;
        LEFT_BUTTON   = 1;
        MIDDLE_BUTTON = 2;
        RIGHT_BUTTON  = 4;
        WHEEL_UP      = 8;
        WHEEL_DOWN    = 16;
    }

    uint32 mask = 1; // Button mask.
    int32 x = 2;     // x position.
    int32 y = 3;     // y position.
}

message ClipboardEvent
{
    string mime_type = 1;
    bytes data = 2;
}

message CursorShape
{
    enum Flags
    {
        UNKNOWN     = 0;
        RESET_CACHE = 64;
        CACHE       = 128;
    }

    // We use only the first byte in this field.
    // If bit 7 is set to 1, then bits 0-4 contain the cursor index in the cache.
    // If bit 7 is not set, then the cursor image is received.
    // If bit 6 is set to 1, then the command to reset the contents of the cache
    // is received, and bits 0-4 contain a new cache size.
    // Bit 5 is reserved.
    uint32 flags = 1;

    // Width, height (in screen pixels) of the cursor.
    int32 width  = 2;
    int32 height = 3;

    // X,Y coordinates (relative to upper-left corner) of the cursor hotspot.
    int32 hotspot_x = 4;
    int32 hotspot_y = 5;

    // Cursor pixmap data in 32-bit BGRA format compressed with ZLIB.
    bytes data = 6;
}

message VideoRect
{
    int32 x = 1;
    int32 y = 2;
    int32 width = 3;
    int32 height = 4;
}

message VideoPixelFormat
{
    uint32 bits_per_pixel = 1;

    uint32 red_max   = 2;
    uint32 green_max = 3;
    uint32 blue_max  = 4;

    uint32 red_shift   = 5;
    uint32 green_shift = 6;
    uint32 blue_shift  = 7;
}

// Identifies how the image was encoded.
enum VideoEncoding
{
    VIDEO_ENCODING_UNKNOWN = 0;
    VIDEO_ENCODING_ZLIB    = 1;
    VIDEO_ENCODING_VP8     = 2;
    VIDEO_ENCODING_VP9     = 4; // LossLess
}

message VideoSize
{
    int32 width = 1;
    int32 height = 2;
}

message VideoPacketFormat
{
    VideoSize screen_size = 1;
    VideoPixelFormat pixel_format = 2;
}

message VideoPacket
{
    VideoEncoding encoding = 1;

    // If the screen size or the pixel format has changed, the field must be filled.
    VideoPacketFormat format = 2;

    // The list of changed rectangles (areas) of the screen.
    repeated VideoRect dirty_rect = 3;

    // Video packet data.
    bytes data = 4;
}

enum AudioEncoding
{
    AUDIO_ENCODING_INVALID = 0;
    AUDIO_ENCODING_OPUS    = 1;
}

message AudioPacket
{
    AudioEncoding encoding = 1;
}

enum DesktopSessionFeatures
{
    FEATURE_NONE         = 0;
    FEATURE_CURSOR_SHAPE = 1;
    FEATURE_CLIPBOARD    = 2;
}

message DesktopSessionConfigRequest
{
    uint32 video_encodings = 1;
    uint32 audio_encodings = 2;
    uint32 features = 3;
}

message DesktopSessionConfig
{
    enum Flags
    {
        ENABLE_NONE             = 0;
        ENABLE_CURSOR_SHAPE     = 1;
        ENABLE_CLIPBOARD        = 2;
        ENABLE_AUDIO            = 3;
    }

    uint32 flags = 1;

    VideoEncoding video_encoding = 2;
    AudioEncoding audio_encoding = 3;

    VideoPixelFormat pixel_format = 4;

    uint32 update_interval = 5;
    uint32 compress_ratio = 6;
}
